### PR TITLE
Use cached testbedInfo and host_vars in inventory in dut selection fixtures

### PR DESCRIPTION
**Pull request by sanmalho-git**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?

pytest_generate_tests dynamically adds enum_* fixtures that select DUTs on which the tests are to be parameterized. In order to select the DUTs on which the tests are to run, we need to get the TestbedInfo and also all variables defined in the inventory file for each DUT.  These operations are time consuming - like creating TestbedInfo taking 1-9 seconds, and getting variables from inventory taking 3-5 seconds.

pytest_generate_tests is called for all the tests/fixtures that needs to run in a pytest session.  This was adding a long time when trying to execute many tests.

This issue was reported in issue #2790 - pytest_generate_tests in tests/conftest.py takes much more time than before

#### How did you do it?

To fix this, we need to create the TestbedInfo only once, store it, and then use the stored value in the next execution of pytest_generate_tests, rather than re-creating TestbedInfo.  Similary, the variables for all the DUTs in the testbed should be read once from the inventory files, stored, and the re-used in the next execution of pytest_generate_tests.

PR #2789 Added caching capability to store any facts using pickle. 
PR #2856 added caching capability for TestbedInfo
PR #2873 added caching capability for variables for DUTs in the inventory files.

We use the cached TestbedInfo and DUT variables in the selection of DUTs in pytest_generate_tests.

#### How did you verify/test it?
Ran tests that use the dut selection fixtures and validated that the delay is seen only once, and not in every call to pytest_generate_tests.

Validated the execution time reduced significantly when using caching when executing iface_namingmode test cases which have ~30 tests.

### CODE REVIEW

**Comment by sanmalho-git** [comment]
retest vsimage please

File: tests/conftest.py
Comment (yxieca) [inline tests/conftest.py:790]: This change is still causing all testcases to try collect testbed_info once, so if a test doesn't need any enumeration, it will still collect the testbed info once?

Would it make sense to put the collector and global variable cache in a helper function and call when needed?

**Comment by sanmalho-git** [comment]
retest vsimage please

**Comment by wangxin** [comment]
@sanmalho-git After some investigation, I feel that the previously added caching function can be used here to avoid global variables. I have enhanced the caching function and enabled caching of TestbedInfo object in PR #2856. 

With the change of PR #2856, the performance of pytest_generate_tests is significantly improved and there is no need to introduce global variables into conftest.py. Well, I am not completely correct about the statement of global variables. My fix still introduced a global variable `cache`. This variable holds an singleton instance of tests.common.cache::FactsCache.

I have plan to enable cache of more information, like host visible variables, then the performance of pytest_generate_tests should can be further improved. I'll submit more PRs to address that.

**Comment by sanmalho-git** [comment]
@wangxin Agreed. Will wait for your PR to be merged and then will modify the code to use the newly introduced caching for the TestbedInfo.

**Comment by yxieca** [comment]
@sanmalho-git can you resolve the merge conflict and ping us for another review?

**Comment by sanmalho-git** [comment]
@yxieca I was waiting for PR# 2873 from Xin to incorporate his changes as well. Once that is merged, will do the rebase and resolve conflicts.

**Comment by wangxin** [comment]
PR #2873 was merged. This PR should be unblocked now.

**Comment by sanmalho-git** [comment]
retest vsimage please

File: tests/conftest.py
Comment (wangxin) [inline tests/conftest.py:630]: If there is neither 'hwsku' nor 'sonic_hwsku', then variable `a_host_hwsku` would be undefined. Line 631 would raise exception. It needs to be initialized in advance.

File: tests/conftest.py
Comment (wangxin) [inline tests/conftest.py:826]: If we re-run function `generate_params_hostname_rand_per_hwsku` every time instead of using a global variable, will the performance of generating tests noticeably degraded?

Comment (sanmalho-git) [reply tests/conftest.py:826]: The issue is not with performance, but the fact that if we have multiple tests within the module and 'module' scoped fixtures, then each fixture/test could pick a different DUT (because of random) and thus inconsistency. So, we need to maintain the randomly selected DUTs per module.

File: tests/conftest.py
Comment (sanmalho-git) [reply tests/conftest.py:631]: Added commit to initialize the variable to None

**Review by wangxin — APPROVED** [review]

**Comment by wangxin** [comment]
@sanmalho-git  Can you update the PR description to match the actual code change? Then we can move forward with merging it.

**Comment by sanmalho-git** [comment]
Changed the title

**Comment by wangxin** [comment]
Also update the content under **What is the motivation for this PR?** and **How did you do it?**? We generally use these information as commit message while merging PRs. The commit message needs to be accurate.

**Pull request merged**
